File: C-Plus-Plus/ds/singly linked list.cpp
/*
All Basics Features of singly Linked List

@author: Rishikeshrajrxl
@created: 16/08/2020
*/


#include <iostream>
using namespace std;

struct node
{                                //declaration of node
int data;
struct node*next;
}*ptr,*head,*tail,*temp;         //pointer to node

void Insert_At_Front();
void Insert_At_End();
void Insert_At_Pos();            //function declaration
void Delete_At_Front();
void Delete_At_End();
void Display();
void Length();
int ele;			//ele is the input data from the user


int main()              //Main Function Begins Here
{
int ch;
cout<< "\n\n\t All Basic Features of Singly Linked List"<<endl;         //Displaying the Features in a tabular manner         
cout<<"\t **************************************************"<<endl;

do
{
cout<<"\t 1) Insert at Front\t \t 2) Insert at End "<<endl; 
cout<<"\t 3) Insert at position\t \t 4) Delete at Front"<<endl;
cout<<"\t 5) Delete at End\t \t 6) Display"<<endl;
cout<<"\t 7) Length\t \t\t 8) Exit"<<endl;
cout<<"\t ***************************************************\n ENTER YOUR CHOICE: "<<endl;
cin >>ch;
  switch(ch)
  {
    case 1:
    Insert_At_Front();
    break;

    case 2:
    Insert_At_End();
    break;

    case 3:
    Insert_At_Pos();
    break;

    case 4:
    Delete_At_Front();
    break;

    case 5:
    Delete_At_End();
    break;

    case 6:
    Display();
    break;

    case 7:
    Length();
    break;

    case 8:
    exit(0);
    break;
    
    default:
    cout<<"Invalid Input. Please enter the Number between 1 to 8 "<<endl;
    
  }
}while(ch!=0);
    return 0;
}
//-------------------------------------------------------------
void Insert_At_End()
{
    ptr=(struct node*)malloc(sizeof(struct node));
    cout<<"ENTER THE ELEMENT"<<endl;
    cin>>ele;
    ptr->data=ele;
    ptr->next=NULL;
if(head==NULL)
 {
    head=ptr;
    tail=ptr;
         cout<<"Inserted Successfully."<<endl;
         cout<<"\t **************************************************"<<endl;
 }
 else
 {
    tail->next=ptr;
    tail=ptr;
          cout<<"Inserted Successfully."<<endl;
          cout<<"\t **************************************************"<<endl; 
 }
}
//--------------------------------------------------------------
void Insert_At_Front()
{
    ptr=(struct node*)malloc(sizeof(struct node));
    cout<<"ENTER THE ELEMENT"<<endl;
    cin>>ele;
    ptr->data=ele;
if(head==NULL)
 {
    head=ptr;
    tail=ptr;
    ptr->next=NULL;
     cout<<"Inserted Successfully."<<endl;
     cout<<"\t **************************************************"<<endl;
 }
else
 {
    ptr->next=head;
    head=ptr;
      cout<<"Inserted Successfully."<<endl;
      cout<<"\t **************************************************"<<endl;

 }
}

//-------------------------------------------------------------
void Insert_At_Pos()
{

    int loc;
    struct node*pos;
    ptr=(struct node*)malloc(sizeof(struct node));
    cout<<"ENTER THE ELEMENT"<<endl;
    cin>>ele;
    ptr->data=ele;
    cout<<"ENTER THE LOCATION"<<endl;
    cin>>loc;
    pos=head;

  for(int i=1;i<loc-1;i++)
  {
    pos=pos->next;
  }
  ptr->next=pos->next;
  pos->next=ptr;
      cout<<"Inserted Successfully."<<endl;
      cout<<"\t *************************************************"<<endl;

}

//-------------------------------------------------------------
void Delete_At_Front()
{
    if(head==NULL)
    cout<<"List is Empty"<<endl;
    else
    {
        ptr=head;
        head=ptr->next;
	   cout<<"Deleted Successfully"<<endl;
	   cout<<"\t **************************************************"<<endl;
        free(ptr);
    }
}

//-------------------------------------------------------------
void Delete_At_End()
{
    if(head==NULL)
    cout<<"List is Empty"<<endl;
    else
    {
        temp=head;
        while(temp->next==NULL)
        {
            ptr=temp;
            temp=ptr->next;
        }
    ptr->next=NULL;
    cout<<"Deleted Successfully"<<endl;
    cout<<"\t **************************************************"<<endl;
    free(temp);
    }
}
//----------------------------------------------------------------------------------
void Display()                         //Traversing throughout the Linked LIst
{
    int i=1;
    temp=head;
    cout<<"S.No\t Data\t Pointer"<<endl;
    while(temp!=NULL)
    {
        cout<<i<<"\t"<<temp->data<<"\t"<<temp->next<<endl;
        temp=temp->next;
        i++;
    }
    cout<<"\t ***********************************************"<<endl;
}
//------------------------------------------------------------------------------------

void Length()
{

    int i,Count=0;
    temp=head;
    while(temp!=NULL)
    {
        Count+=1;
        temp=temp->next;
        i++;
    }
    cout<<"Total no. of Nodes : "<<Count<<endl;
    cout<<"\t ***********************************************"<<endl;
}

File: Go/ds/linked_list/singly_linked_list/singly_linked_list.go
/*
	==============================================================
		 Implementing Singly Linked List  data structure
	==============================================================
*/

package main

import (
	"fmt"
)

// Node it is user-defined data type used to store data and the address of next node
type Node struct {
	next *Node
	data int
}

// List it is user-defined data type used to represent the linked list
type List struct {
	head *Node
	tail *Node
	size int
}

// initList(): it will initalize the List's fields for first time
func (list *List) initList() {
	list.head = nil
	list.tail = nil
	list.size = 0
}

// isEmpty(): if the head of list is not poiting any node then it will return true otherwise false
func (list *List) isEmpty() bool {
	return list.head == nil
}

// insertAtFront(): if the list is empty then it will assign the same value to head and tail of linked list
// otherwise it will insert node before the head in linked list
func (list *List) insertAtFront() {
	var value int
	fmt.Println("enter the value of element to be inserted at front:")
	fmt.Scanf("%d", &value)
	var node Node
	node.data = value
	if list.isEmpty() {
		list.head = &node
		list.tail = &node
		list.size = 1
	} else {
		node.next = list.head
		list.head = &node
		list.size++
	}
	fmt.Println("value added successfully at the front of the linked list")
}

// insertAtEnd(): if the list is empty then it will assign the same value to head and tail of linked list
// otherwise it will insert node after the tail in linked list
func (list *List) insertAtEnd() {
	var value int
	fmt.Println("enter the value of element to be inserted at the end:")
	fmt.Scanf("%d", &value)
	var node Node
	node.data = value
	if list.isEmpty() {
		list.head = &node
		list.tail = &node
		list.size = 1
	} else {
		list.tail.next = &node
		list.tail = &node
		list.size++
	}
	fmt.Println("value added successfully at the end of the linked list")
}

// insertAtGivenPosition(): if the list is empty then it will assign the same value to head and tail of linked list
// otherwise it will insert node at the given psition in linked list
func (list *List) insertAtGivenPosition() {
	var position int
	var value int
	var node Node

	fmt.Println("enter the position for element")
	fmt.Scanf("%d", &position)
	fmt.Println("enter the value of element to be inserted at above position:")
	fmt.Scanf("%d", &value)
	node.data = value
	if list.isEmpty() {
		list.head = &node
		list.tail = &node
		list.size = 1
	} else {
		temp := list.head
		if position < 2 {
			node.next = list.head
			list.head = &node
		} else if position == list.size+1 {
			list.tail.next = &node
			list.tail = &node
		} else {
			for i := 0; i < position-2; i++ {
				temp = temp.next
			}
			node.next = temp.next
			temp.next = &node
		}
		list.size++

	}
	fmt.Printf("value added successfully at position: %d of the linked\n", position)
}

// we dont need deallocate memory from removed node manually
// because GO garbage collector will do it by it self automatically
func (list *List) deleteFromFront() {
	if list.isEmpty() {
		fmt.Println("Linked List empty could'n delete element")
	} else {
		list.head = list.head.next
		list.size--
		fmt.Println("successfully deleted element from the front")
	}
}

// deleteFromEnd(): it traverse through all liked list and just dereffrence the last node
func (list *List) deleteFromEnd() {
	if list.isEmpty() {
		fmt.Println("Linked List empty could'n delete element")
	} else {
		temp := list.head
		// iterate unitll temp become second last node
		for i := 0; i < list.size-1; i++ {
			temp = temp.next
		}

		// making second last node's next nil so GC will automatically free the
		// memory used by last node
		temp.next = nil
		list.size--
		fmt.Println("successfully deleted element from the end")
	}
}

func (list *List) deleteFromGivinPosition() {
	if list.isEmpty() {
		fmt.Println("Linked List empty could'n delete element")
	} else {
		temp := list.head
		var position int
		fmt.Println("enter the position of element")
		fmt.Scanf("%d", &position)

		// first node before given position
		for i := 0; i < position-2; i++ {
			temp = temp.next
		}

		// first node after given position
		nextNode := temp.next.next

		// will link the node which come after given position to
		// node which come before given position
		temp.next = nextNode

		// GC will deallocate memory from unsed node
		list.size--
		fmt.Printf("successfully deleted element from the given position: %d\n", position)
	}
}

// length(): will print the total number of node available at a time
func (list *List) length() {
	if list.isEmpty() {
		fmt.Println("Linked List is empty")
	} else {
		fmt.Printf("the number of avialbale node in Linked List is %d\n", list.size)
	}
}

// print(): it will travers through linked list and prints the data of each node
func (list *List) print() {
	if list.isEmpty() {
		fmt.Println("Linked List is empty")
	} else {
		temp := list.head
		fmt.Println("the current state of linked list is:")
		for i := 0; i < list.size; i++ {
			fmt.Printf("%d ", temp.data)
			temp = temp.next
		}
		fmt.Println()
	}
}

func main() {
	var linkList List
	linkList.initList() // initalizing the linkList for first time

	var choice int
	var isContinue bool = true
	fmt.Println("Please enter option")
	for isContinue {
		fmt.Println("1 -> insert element at front\n2 -> insert element at end\n" +
			"3 -> insert element at given position\n4 -> print length\n" +
			"5 -> delete element from front\n6 -> delete element from end\n" +
			"7 -> delete element from given position\n8 -> print the linked list\n" +
			"9 -> exit program")
		fmt.Scanf("%d", &choice)

		switch choice {
		case 1:
			linkList.insertAtFront()
			break
		case 2:
			linkList.insertAtEnd()
			break
		case 3:
			linkList.insertAtGivenPosition()
			break
		case 4:
			linkList.length()
			break
		case 5:
			linkList.deleteFromFront()
			break
		case 6:
			linkList.deleteFromEnd()
			break
		case 7:
			linkList.deleteFromGivinPosition()
			break
		case 8:
			linkList.print()
			break
		case 9:
			isContinue = false
			break
		default:
			fmt.Println("Please enter correct option")
			break
		}
	}
}

/*
	input/output sample

	Please enter option
	1 -> insert element at front
	2 -> insert element at end
	3 -> insert element at given position
	4 -> print length
	5 -> delete element from front
	6 -> delete element from end
	7 -> delete element from given position
	8 -> print the linked list
	9 -> exit program
	1
	enter the value of element to be inserted at front:
	1
	value added successfully at the front of the linked list
	1 -> insert element at front
	2 -> insert element at end
	3 -> insert element at given position
	4 -> print length
	5 -> delete element from front
	6 -> delete element from end
	7 -> delete element from given position
	8 -> print the linked list
	9 -> exit program
	2
	enter the value of element to be inserted at the end:
	2
	value added successfully at the end of the linked list
	1 -> insert element at front
	2 -> insert element at end
	3 -> insert element at given position
	4 -> print length
	5 -> delete element from front
	6 -> delete element from end
	7 -> delete element from given position
	8 -> print the linked list
	9 -> exit program
	3
	enter the position for element
	3
	enter the value of element to be inserted at above position:
	3
	value added successfully at position: 3 of the linked
	1 -> insert element at front
	2 -> insert element at end
	3 -> insert element at given position
	4 -> print length
	5 -> delete element from front
	6 -> delete element from end
	7 -> delete element from given position
	8 -> print the linked list
	9 -> exit program
	8
	the current state of linked list is:
	1 2 3

*/
